bundles/org.openhab.binding.salus/src/main/java/org/openhab/binding/salus/internal/aws/http/AuthenticationHelper.java

```diff
@@ -0,0 +1,436 @@
+            + "E39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9"//
+            + "DE2BCBF6955817183995497CEA956AE515D2261898FA0510"//
+            + "15728E5A8AAAC42DAD33170D04507A33A85521ABDF1CBA64"//
+            + "ECFB850458DBEF0A8AEA71575D060C7DB3970F85A6E1E4C7"//
+            + "ABF5AE8CDB0933D71E8C94E04A25619DCEE3D2261AD2EE6B"//
+            + "F12FFA06D98A0864D87602733EC86A64521F2B18177B200C"//
+            + "BBE117577A615D6C770988C0BAD946E208E24FA074E5AB31"//
+            + "43DB5BFCE0FD108E4B82D120A93AD2CAFFFFFFFFFFFFFFFF";
+    private static final BigInteger N = new BigInteger(HEX_N, 16);
+    private static final BigInteger g = BigInteger.valueOf(2);
+    private static final BigInteger k;
+    private static final int EPHEMERAL_KEY_LENGTH = 1024;
+    private static final int DERIVED_KEY_SIZE = 16;
+    private static final String DERIVED_KEY_INFO = "Caldera Derived Key";
+    private static final ThreadLocal<MessageDigest> THREAD_MESSAGE_DIGEST = ThreadLocal.withInitial(() -> {
+        try {
+            return MessageDigest.getInstance("SHA-256");
+        } catch (NoSuchAlgorithmException e) {
+            throw new SecurityException("Exception in authentication", e);
+        }
+    });
+    private static final SecureRandom SECURE_RANDOM;
+    private static final String ALGORITHM = "HmacSHA256";
+
+    static {
+        try {
+            SECURE_RANDOM = SecureRandom.getInstance("SHA1PRNG");
+
+            MessageDigest messageDigest = THREAD_MESSAGE_DIGEST.get();
+            messageDigest.reset();
+            messageDigest.update(N.toByteArray());
+            byte[] digest = messageDigest.digest(g.toByteArray());
+            k = new BigInteger(1, digest);
+        } catch (NoSuchAlgorithmException e) {
+            throw new SecurityException(e.getMessage(), e);
+        }
+    }
+
+    private BigInteger a;
+    private BigInteger A;
+    private final String userPoolID;
+    private final String clientId;
+    private final String region;
+
+    AuthenticationHelper(String userPoolID, String clientid, String region) {
+        do {
+            a = new BigInteger(EPHEMERAL_KEY_LENGTH, SECURE_RANDOM).mod(N);
+            A = g.modPow(a, N);
+        } while (A.mod(N).equals(BigInteger.ZERO));
+
+        this.userPoolID = userPoolID;
+        this.clientId = clientid;
+        this.region = region;
+    }
+
+    private byte[] getPasswordAuthenticationKey(String userId, byte[] userPassword, BigInteger B, BigInteger salt)
+            throws ShortBufferException, NoSuchAlgorithmException, InvalidKeyException {
+        // Authenticate the password
+        // u = H(A, B)
+        var messageDigest = THREAD_MESSAGE_DIGEST.get();
+        messageDigest.reset();
+        messageDigest.update(A.toByteArray());
+        var u = new BigInteger(1, messageDigest.digest(B.toByteArray()));
+        if (u.equals(BigInteger.ZERO)) {
+            throw new SecurityException("Hash of A and B cannot be zero");
```

Comment:
Not in the list of Exceptions at method decleration.

Reply:
done